src/wasm.rs

```diff
@@ -0,0 +1,130 @@
+use std::{
+    borrow::Cow,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use futures::{stream::SplitSink, Sink, Stream, StreamExt};
+use pharos::{Observable, ObserveConfig};
+use pin_project_lite::pin_project;
+use ws_stream_wasm::{WsErr, WsEvent, WsMessage, WsMeta, WsStream};
+
+use crate::websockets::WebsocketMessage;
+
+/// Creates a new pair of sink and stream which operate on [`WasmWebsocketMessage`] instead of `WsMessage` and `WsEvent` separately.
+pub async fn wasm_websocket_combined_split(
+    mut ws_meta: WsMeta,
+    ws_stream: WsStream,
+) -> (
+    impl Sink<WasmWebsocketMessage, Error = WsErr>,
+    impl Stream<Item = Result<WasmWebsocketMessage, WsErr>>,
+) {
+    let event_stream = ws_meta.observe(ObserveConfig::default()).await.unwrap();
+
+    let (sink, message_stream) = ws_stream.split();
+
+    let message_stream = message_stream.map(|f| Ok::<_, WsErr>(WasmWebsocketMessage::WsMessage(f)));
+
+    let event_stream = event_stream.map(|f| match f {
+        WsEvent::WsErr(e) => Err(e),
+        e => Ok(WasmWebsocketMessage::WsEvent(e)),
+    });
+
+    let stream = futures::stream::select(event_stream, message_stream);
+
+    let sink = FusedWasmWebsocketSink::new(sink);
+
+    (sink, stream)
+}
+
+#[derive(Debug)]
+/// A WebSocket event abstraction to combine the differentiated WebSocket events and messages of `ws_stream_wasm`.
+pub enum WasmWebsocketMessage {
+    /// A WebSocket message, text or binary, was received.
+    WsMessage(ws_stream_wasm::WsMessage),
+    /// A WebSocket event other than a message was received.
+    WsEvent(ws_stream_wasm::WsEvent),
+}
+
+impl WebsocketMessage for WasmWebsocketMessage {
```

Comment:
We should either keep the messages implementation together in the same file or split the tungstenite implementation in its own file. The goal is to be consistent.

Reply:
Done